In [ ]:
import numpy as np
import pandas as pd

1. Сгенерировать случайным образом массив numpy из 1000 значений
нормально распределенной случайной величины (numpy.random.normal) с
мат.
ожиданием
M = 1.0
и стандартным отклонением
s = 1.0.
Преобразовать его в объект Series.

In [ ]:
normal_array = np.random.normal(loc=1.0, scale=1.0, size=1000)
normal_series = pd.Series(normal_array)
print(normal_series)

0      0.084953
1      1.626895
2      2.149213
3     -0.086395
4      0.690301
         ...   
995    1.174183
996    2.292681
997    1.479288
998   -0.195393
999   -1.798188
Length: 1000, dtype: float64


2. Вычислить, какая доля всех значений находится в диапазоне (M-s; M+s).

In [ ]:
sigma_range = ((normal_series > 0.0) & (normal_series < 2.0)).mean()
print(sigma_range)

0.703


3. Вычислить, какая доля всех значений находится в диапазоне (M-3s; M+3s).
Какой должна быть эта доля теоретически? Насколько реальный
результат соответствует теории?

In [ ]:
sigma3_range = ((normal_series > -2.0) & (normal_series < 4.0)).mean()
print(sigma3_range)

0.998


Обычно в пределах одного среднеквадратического отклонения лежит 68,26% значений, мы получили очень приближенную к этому значению долю. В пределах трех — 99,72%, у нас эьл 0.998, тоже очень близкое значение, значит, массив действительно состоит из значений нормально распределенной случайной величины.

4. Заменить каждое значение x в серии на его квадратный корень. Результат записать в новый объект Series. Почему
возникает предупреждение, и что происходит с теми значениями, для
которых возникает предупреждение?

In [ ]:
series_sqrt = np.sqrt(normal_series)
print(series_sqrt)

0      0.291467
1      1.275498
2      1.466019
3           NaN
4      0.830844
         ...   
995    1.083597
996    1.514160
997    1.216260
998         NaN
999         NaN
Length: 1000, dtype: float64


/usr/local/lib/python3.12/dist-packages/pandas/core/arraylike.py:399: RuntimeWarning: invalid value encountered in sqrt
  result = getattr(ufunc, method)(*inputs, **kwargs)


Предупреждение возникает из-за отрицательных значений, для них вместо корня записываетсяч NaN

5. Посчитать среднее арифметическое для получившихся значений.
Отсутствующие значения (NaN) учитываться не должны.

In [ ]:
average_value = series_sqrt.mean()
print(average_value)

1.057911157725125


6. На основе двух объектов Series
создать DataFrame с двумя столбцами. Названия (явные индексы) для
столбцов: «number» и «root» соответственно. Явные индексы для строк не
задавать. Вывести первые 6 строк из созданного датафрейма.

In [ ]:
df = pd.DataFrame({'number': normal_series, 'root': series_sqrt})
print(df.head(6))

     number      root
0  0.084953  0.291467
1  1.626895  1.275498
2  2.149213  1.466019
3 -0.086395       NaN
4  0.690301  0.830844
5 -0.178305       NaN


7. С помощью функции query найти в датафрейме записи, в которых
значение квадратного корня находится в диапазоне от 1.8 до 1.9. Вывести
результат.

In [ ]:
filtered = df.query('1.8 <= root <= 1.9')
print(filtered)

       number      root
27   3.374872  1.837082
94   3.515939  1.875084
108  3.377874  1.837899
669  3.351235  1.830638
715  3.439486  1.854585
746  3.348145  1.829794
801  3.490547  1.868301
890  3.384820  1.839788
934  3.274604  1.809587
